src/Symfony/Component/TypeInfo/Type.php

```diff
@@ -25,35 +24,70 @@ abstract class Type implements \Stringable
 {
     use TypeFactoryTrait;
 
-    abstract public function getBaseType(): BuiltinType|ObjectType;
-
     /**
-     * @param TypeIdentifier|class-string $subject
+     * Traverses and yields the whole type inheritance graph.
+     *
+     * For example traversing "string|\Traversable<int>" will yield
+     * - string
+     * - int
+     * - \Traversable<int>
+     * - string|\Traversable<int>
+     *
+     * @return iterable<self>
      */
-    abstract public function isA(TypeIdentifier|string $subject): bool;
+    public function traverse(): iterable
+    {
+        if ($this instanceof CompositeTypeInterface) {
+            foreach ($this->getTypes() as $type) {
+                yield from $type->traverse();
+            }
+        }
+
+        if ($this instanceof WrappingTypeInterface) {
+            yield from $this->getWrappedType()->traverse();
+        }
 
-    abstract public function asNonNullable(): self;
+        yield $this;
+    }
 
     /**
-     * @param callable(Type): bool $callable
+     * Tells if the type or one of its component (@see traverse()) is what is expected.
+     *
+     * @param TypeIdentifier|class-string $expected
      */
-    public function is(callable $callable): bool
+    public function isA(TypeIdentifier|string $expected): bool
     {
-        return $callable($this);
+        foreach ($this->traverse() as $type) {
+            if ($type !== $this && $type->isA($expected)) {
+                return true;
+            }
+        }
+
+        return false;
     }
 
+    /**
+     * Tells if the type or one of its component (@see traverse()) is nullable.
+     *
+     * @param TypeIdentifier|class-string $expected
+     */
     public function isNullable(): bool
     {
-        return $this->is(fn (Type $t): bool => $t->isA(TypeIdentifier::NULL) || $t->isA(TypeIdentifier::MIXED));
+        return $this->isA(TypeIdentifier::NULL) || $this->isA(TypeIdentifier::MIXED);
     }
 
     /**
-     * Graceful fallback for unexisting methods.
+     * Tells if the type or one of its component (@see traverse()) is a scalar.
      *
-     * @param list<mixed> $arguments
+     * @param TypeIdentifier|class-string $expected
      */
-    public function __call(string $method, array $arguments): mixed
+    public function isScalar(): bool
     {
-        throw new LogicException(\sprintf('Cannot call "%s" on "%s" type.', $method, $this));
+        return $this->isA(TypeIdentifier::INT)
```

Comment:
instead of separate `isA` calls that will require separate traversals, we should perform a single traversing using a combined check for all identifiers (maybe with an API inspired by `is` taking a callback with a type)

Reply:
Great idea yes, I updated the method signature so that it can take a variadic argument